packages/core/src/elasticStorage.ts

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright © 2020 Anticrm Platform Contributors.
+// 
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// 
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import type { Storage } from './storage'
+import { TxProcessor } from './storage'
+import type { Emb, Doc, Ref, Class, Data, Collection } from './classes'
+import type { TxAddCollection, TxCreateObject } from './tx'
+import type { Hierarchy } from './hierarchy'
+import { Client, RequestParams } from '@elastic/elasticsearch'
+import { generateId } from './utils'
+
+export type ConnectionParams = {
+  url: string,
+  username: string,
+  password: string
+}
+
+export class ElasticStorage extends TxProcessor implements Storage {
+  private readonly hierarchy: Hierarchy
+  private readonly client: Client
+  private readonly workspace: string
+
+  constructor (hierarchy: Hierarchy, workspace: string, connection: ConnectionParams) {
+    super()
+    this.hierarchy = hierarchy
+    this.workspace = workspace ?? 'workspace'
```

Comment:
It is not clear what workspace param passed could be undefined, please use ? or explicitly specify type | undefined in definition.